client/noxfile.py

```diff
@@ -53,6 +57,11 @@ def type_check(session):
 def tests(session):
     """Execute tests."""
     with tempfile.NamedTemporaryFile() as temp_file:
+        # TODO: mjquinn asks: why do we need these `session.install`
+        # invocations? CONTRIBUTING.md already tells developers to run
+        # `poetry install`. If I run `nox -s tests` with these lines enabled,
+        # I get `Error: python is not installed into the virtualenv`. If I
+        # comment them out, I can run the tests as expected.
```

Comment:
Full log is below.

I have python 3.10 but nothing else. Makes sense what you say about sessions/virtualenvs, I just would expect it to pick up the python binary in the virtualenv that it explicitly lists in the error message. I tried putting `external=True` everywhere like it suggests, but nothing changed. 

```
(bqms-run-py3.10) mjquinn@mjquinn:~/dwh-migration-tools/client$ nox -s tests --verbose -- unit pytest_verbose
nox > Running session tests-3.7
nox > Missing interpreters will error by default on CI systems.
nox > Session tests-3.7 skipped: Python interpreter 3.7 not found.
nox > Running session tests-3.8
nox > Missing interpreters will error by default on CI systems.
nox > Session tests-3.8 skipped: Python interpreter 3.8 not found.
nox > Running session tests-3.9
nox > Missing interpreters will error by default on CI systems.
nox > Session tests-3.9 skipped: Python interpreter 3.9 not found.
nox > Running session tests-3.10
nox > Creating virtual environment (virtualenv) using python3.10 in .nox/tests-3-10
created virtual environment CPython3.10.9.final.0-64 in 239ms
  creator CPython3Posix(dest=/usr/local/google/home/mjquinn/dwh-migration-tools/client/.nox/tests-3-10, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=/usr/local/google/home/mjquinn/.local/share/virtualenv)
    added seed packages: pip==23.0, setuptools==67.1.0, wheel==0.38.4
  activators BashActivator,CShellActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator

nox > python -m pip install poetry
nox > Error: python is not installed into the virtualenv, it is located at /usr/local/google/home/mjquinn/.cache/pypoetry/virtualenvs/bqms-run-cje5m_1w-py3.10/bin/python. Pass external=True into run() to explicitly allow this.
nox > Session tests-3.10 failed.
nox > Ran multiple sessions:
nox > * tests-3.7: skipped
nox > * tests-3.8: skipped
nox > * tests-3.9: skipped
nox > * tests-3.10: failed
```

Reply:
> I just would expect it to pick up the python binary in the virtualenv that it explicitly lists in the error message

The virtualenv listed in the error message (`/usr/local/google/home/mjquinn/.cache/pypoetry/virtualenvs/bqms-run-cje5m_1w-py3.10`) is the virtualenv that is created by `poetry install` that you would use during development to run the tool manually. It's also the virtualenv that you would want your IDE to reference. Finally, it's the virtualenv that is used for formatting, linting and type checking because [we are passing python=false to those nox sessions](https://github.com/google/dwh-migration-tools/blob/main/client/noxfile.py#L45) ([docs](python=False)). Essentially, it's the persistent virtualenv that you use during development.

The virtualenvs that nox creates are located at `/usr/local/google/home/mjquinn/dwh-migration-tools/client/.nox`. They are used for tests because they are supposed to be transient and isolated from any virtualenv you use for development. On my machine, there is one created for each Python interpreter minor version we are testing against:

```shell
> ll .nox
total 16
drwxr-x--- 4 ajwelch primarygroup 4096 Feb  2 14:42 tests-3-10
drwxr-x--- 4 ajwelch primarygroup 4096 Feb  2 14:35 tests-3-7
drwxr-x--- 4 ajwelch primarygroup 4096 Feb  2 14:38 tests-3-8
drwxr-x--- 4 ajwelch primarygroup 4096 Feb  2 14:40 tests-3-9
```

Inside each of these nox virtualenvs, there should be a symlink to the Python interpreter. For example, on my machine the nox virtualenv for Python 3.10 symlinks to the Python 3.10 interpreter I have installed on my machine via [pyenv](https://github.com/pyenv/pyenv):
```shell
> ll .nox/tests-3-10/bin/python
lrwxrwxrwx 1 ajwelch primarygroup 68 Feb  2 14:42 .nox/tests-3-10/bin/python -> /usr/local/google/home/ajwelch/.pyenv/versions/3.10.1/bin/python3.10
```

 I think something odd is going on during your nox virtualenv creation. I'm curious what is in `usr/local/google/home/mjquinn/dwh-migration-tools/client/.nox/tests-3-10/bin` on your machine? The error message leads me to believe there is no `python` symlink there.

>  I tried putting external=True everywhere like it suggests, but nothing changed.

Ideally, you should not have to pass `external=True` as this is basically telling nox not to worry about isolation and to go ahead and use the python interpreter from your poetry dev virtualenv.

Reply:
> I'm curious what is in usr/local/google/home/mjquinn/dwh-migration-tools/client/.nox/tests-3-10/bin on your machine?

You're right, there is no python symlink, there isn't even a `bin/` subdirectory:

```bash
(bqms-run-py3.10) mjquinn@mjquinn:~/dwh-migration-tools/client$ ll .nox/tests-3-10/
total 24
drwxr-xr-x 4 mjquinn primarygroup 4096 Feb 16 17:14 ./
drwxr-xr-x 3 mjquinn primarygroup 4096 Feb 16 17:14 ../
-rw-r--r-- 1 mjquinn primarygroup   40 Feb 16 17:14 .gitignore
drwxr-xr-x 3 mjquinn primarygroup 4096 Feb 16 17:14 lib/
drwxr-xr-x 4 mjquinn primarygroup 4096 Feb 16 17:14 local/
-rw-r--r-- 1 mjquinn primarygroup  210 Feb 16 17:14 pyvenv.cfg
```

> Ideally, you should not have to pass external=True as this is basically telling nox not to worry about isolation

Got it, I figured it wasn't a permanent change that should be made, but just gave it a go to see if anything changed for debugging purposes.

In the past when creating virtualenvs I've done so manually, this is my first time using a higher-level tool to create them. Since I'm running the same commands you are (in the README) it seems to me like this isn't a poetry issue; wonder if me running on a CloudTop has anything to do with it...

Reply:
> wonder if me running on a CloudTop has anything to do with it...

I'm running on a CloudTop as well. I just spun up a fresh CloudTop that has never been used before. Here are the commands from my `.zsh_history` that were used to install the prerequisites mentioned in `CONTRIBUTING.md`:

```shell
: 1677451020:0;mkdir code
: 1677451226:0;cd code
: 1677451258:0;git clone https://github.com/google/dwh-migration-tools.git
: 1677451359:0;curl -sSL https://install.python-poetry.org | python3 -
: 1677451415:0;nano ~/.zshrc
: 1677451473:0;sudo apt install -y nodejs
: 1677451516:0;node --version
: 1677451531:0;sudo apt install -y npm
: 1677451614:0;npm --version
: 1677451619:0;cd /opt
: 1677451624:0;curl -O https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-cli-419.0.0-linux-x86_64.tar.gz
: 1677451694:0;tar -xf google-cloud-cli-419.0.0-linux-x86_64.tar.gz
: 1677451733:0;./google-cloud-sdk/install.sh
: 1677451779:0;gcloud --help
: 1677451795:0;gcloud init
: 1677452010:0;go install github.com/google/addlicense@latest
: 1677452035:0;cd ~/code/dwh-migration-tools/client
: 1677452042:0;poetry install
: 1677452165:0;sudo apt install -y pkg-config
: 1677452180:0;poetry install
: 1677452262:0;poetry shell
```

I'm then able to run the unit tests:

```shell
ajwelch@ajwtest ~/code/dwh-migration-tools/client
 % nox -s tests -- unit
nox > Running session tests-3.7
nox > Missing interpreters will error by default on CI systems.
nox > Session tests-3.7 skipped: Python interpreter 3.7 not found.
nox > Running session tests-3.8
nox > Missing interpreters will error by default on CI systems.
nox > Session tests-3.8 skipped: Python interpreter 3.8 not found.
nox > Running session tests-3.9
nox > Missing interpreters will error by default on CI systems.
nox > Session tests-3.9 skipped: Python interpreter 3.9 not found.
nox > Running session tests-3.10
nox > Creating virtual environment (virtualenv) using python3.10 in .nox/tests-3-10
nox > python -m pip install poetry
nox > poetry export --with dev --without-hashes --output /tmp/tmp7ufylmpl
nox > python -m pip install -r /tmp/tmp7ufylmpl
nox > python -m pip install .
nox > coverage run --source=bqms_run -m pytest tests/unit
================================================= test session starts =================================================
platform linux -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0
rootdir: /usr/local/google/home/ajwelch/code/dwh-migration-tools/client, configfile: pyproject.toml
collected 17 items                                                                                                    

tests/unit/test_ksh.py ..........                                                                               [ 58%]
tests/unit/test_macros.py .......                                                                               [100%]

================================================== warnings summary ===================================================
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
  /usr/local/google/home/ajwelch/code/dwh-migration-tools/client/.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
  /usr/local/google/home/ajwelch/code/dwh-migration-tools/client/.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.cloud')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2309
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2309
.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2309
  /usr/local/google/home/ajwelch/code/dwh-migration-tools/client/.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2309: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(parent)

.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804
  /usr/local/google/home/ajwelch/code/dwh-migration-tools/client/.nox/tests-3-10/lib/python3.10/site-packages/pkg_resources/__init__.py:2804: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.logging')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

.nox/tests-3-10/lib/python3.10/site-packages/google/rpc/__init__.py:20
  /usr/local/google/home/ajwelch/code/dwh-migration-tools/client/.nox/tests-3-10/lib/python3.10/site-packages/google/rpc/__init__.py:20: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.rpc')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    pkg_resources.declare_namespace(__name__)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================================== 17 passed, 15 warnings in 0.71s ===========================================
nox > coverage report -m
Name                                       Stmts   Miss  Cover   Missing
------------------------------------------------------------------------
bqms_run/__init__.py                           1      0   100%
bqms_run/encoding.py                          14      1    93%   34
bqms_run/gcp/__init__.py                       1      1     0%   14
bqms_run/gcp/bqms/__init__.py                  1      1     0%   14
bqms_run/gcp/bqms/object_name_mapping.py      36     36     0%   14-98
bqms_run/gcp/bqms/request.py                  33     33     0%   14-130
bqms_run/gcp/bqms/source_env.py               14     14     0%   14-45
bqms_run/gcp/bqms/translation_type.py         23     23     0%   14-191
bqms_run/gcp/gcs.py                           66     66     0%   14-183
bqms_run/hooks.py                             42     21    50%   57, 62, 70-73, 82-104, 120, 125
bqms_run/ksh.py                               91      4    96%   45, 148, 165, 227
bqms_run/macros.py                           118      7    94%   39, 46-47, 98, 101, 119, 199
bqms_run/main.py                             175    175     0%   14-323
bqms_run/paths.py                             96     48    50%   47-53, 57-63, 67-72, 76-80, 84-89, 93-98, 102, 158-159, 162-163, 171-175, 180-184
bqms_run/workflow.py                          76     76     0%   14-187
------------------------------------------------------------------------
TOTAL                                        787    506    36%
nox > Session tests-3.10 was successful.
nox > Ran multiple sessions:
nox > * tests-3.7: skipped
nox > * tests-3.8: skipped
nox > * tests-3.9: skipped
nox > * tests-3.10: success
```

And I can see that the nox virtualenv is symlinked to the system Python 3 interpreter (which is 3.10 and thus tests for 3.7 to 3.9 were skipped):
```shell
ajwelch@ajwtest ~/code/dwh-migration-tools/client
 % ls -la .nox/tests-3-10/bin/python
lrwxrwxrwx 1 ajwelch primarygroup 19 Feb 26 22:57 .nox/tests-3-10/bin/python -> /usr/bin/python3.10
```

I think you might have to just try blowing things away and reinstalling or perhaps using `pyenv` to manage Python interpreters instead of defaulting to the system Python.

Reply:
Ok, will give that a go, appreciate the advice.